### PR TITLE
RankLLM Revamp - VLLM support, APEER prompt, fixed caching, improved documentation

### PR DESCRIPTION
**Pull request by ronakice**
# Pull Request Checklist

## Reference Issue

Please provide the reference to issue this PR is addressing (# followed by the issue number). If there is no associated issue, write "N/A".

ref: 

## Checklist Items

Before submitting your pull request, please review these items:

- [x] Have you followed the [contributing guidelines](CONTRIBUTING.md)?
- [x] Have you verified that there are no existing [Pull Requests](https://github.com/castorini/rank_llm/pulls) for the same update/change?
- [x] Have you updated any relevant documentation or added new tests where needed?

# PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... 
    - Description:

### CODE REVIEW

File: src/rank_llm/retrieve/retriever.py [hunk 1]
```diff
@@ -145,18 +145,17 @@ def retrieve(
...
         if self._retrieval_mode == RetrievalMode.DATASET:
             candidates_file = Path(
-                f"{retrieve_results_dirname}/{self._retrieval_method.name}/retrieve_results_{self._dataset}_top{k}.json"
+                f"{retrieve_results_dirname}/{self._retrieval_method.name}/retrieve_results_{self._dataset}_top{k}.jsonl"
             )
-            query_name = f"{self._retrieval_method.name}/retrieve_results_{self._dataset}_top{k}.json"
+            query_name = f"{self._retrieval_method.name}/retrieve_results_{self._dataset}_top{k}.jsonl"
             if not candidates_file.is_file():
                 try:
                     # TODO: Fix caching
```
Comment (sahel-sh) [inline src/rank_llm/retrieve/retriever.py:153]: please remove the todo

File: src/rank_llm/retrieve/retriever.py [hunk 2]
```diff
@@ -169,10 +168,9 @@ def retrieve(
                 # if HITS_INFO[query_name]["md5"] != md5_local:
```
Comment (sahel-sh) [inline src/rank_llm/retrieve/retriever.py:168]: does this need fixing for cache use?

File: src/rank_llm/rerank/rank_listwise_os_llm.py [hunk 3]
```diff
@@ -71,6 +81,19 @@ def __init__(
...
                 self._examples = list(json_file)[1:-1]
 
+    def run_llm_batched(self, prompts: List[str | List[Dict[str, str]]],
+                        current_window_size: Optional[int] = None) -> List[Tuple[str, int]]:
+        sampling_params = SamplingParams(
+                temperature=0.0,
+                max_tokens=self.num_output_tokens(current_window_size),
+                min_tokens=self.num_output_tokens(current_window_size),
+            )
+        outputs = self._llm.generate(prompts, sampling_params)
```
Comment (sahel-sh) [inline src/rank_llm/rerank/rank_listwise_os_llm.py:91]: do we need to specify the batch size or make it configurable? how is it defined/controlled?

File: src/rank_llm/scripts/run_rank_llm.py [hunk 4]
```diff
@@ -153,5 +155,10 @@ def main(args):
         default="You are RankLLM, an intelligent assistant that can rank passages based on their relevancy to the query.",
         help="the system message used in prompts",
     )
+    parser.add_argument(
+        "--batched",
```
Comment (sahel-sh) [inline src/rank_llm/scripts/run_rank_llm.py:159]: This batch can be confused with rerank_batch, the latter is for reranking multiple queries regardless of if it is happening one query at a time or multiple queries at once. The former refers to reranking multiple queries at once. from the default param values it looks like we want to support rerank_batch function with batch=false? if yes, I think the two different meanings of batch can be confusing, maybe we need to rename one of the two usecases? WDYT @ronakice , @lintool

File: pyproject.toml [hunk 5]
```diff
@@ -26,6 +26,11 @@ requires-python = ">= 3.10"
 [tool.setuptools.dynamic]
 dependencies = {file = ["requirements.txt"]}
 
+[project.optional-dependencies]
+vllm = [
```
Comment (sahel-sh) [inline pyproject.toml:30]: hm, what would happen to imports from vllm if this optional dependency is not installed? why is this optional?

Comment (ronakice) [reply pyproject.toml:30]: I feel optional is better, I don't want to add more dependencies unless needed.

File: src/rank_llm/rerank/rank_listwise_os_llm.py (hunk 3 above)
Comment (ronakice) [reply src/rank_llm/rerank/rank_listwise_os_llm.py:107]: VLLM does auto-batching based on GPU size, so the batch size is dynamically allocated I believe

File: src/rank_llm/scripts/run_rank_llm.py (hunk 4 above)
Comment (ronakice) [reply src/rank_llm/scripts/run_rank_llm.py:159]: Hmm i agree this could be a source of confusion for users. Batching usually means multiple queries being processed at once correct? In which case `rerank_batch` needs renaming?

File: pyproject.toml (hunk 5 above)
Comment (sahel-sh) [reply pyproject.toml:30]: this is good, thanks

File: src/rank_llm/scripts/run_rank_llm.py (hunk 4 above)
Comment (sahel-sh) [reply src/rank_llm/scripts/run_rank_llm.py:159]: with my understanding batch processing is async and yes it is normally multiple queries at once, but it doesn't restrict the batch size to be strictly greater than 1. IIRC, @lintool  chose `rerank` and `rerank_batch` to align with pyserini function names (`retrieve` and `retrieve_batch`).
@ronakice WDYT about changing `--batched` to `--vllm-batched`?
 I propose to have an `--inference-method` enum (or some similar name) with FastChat as the default value. The user can set it to vllm if they want to use vllm rather than fastchat inference.

Comment (ronakice) [reply src/rank_llm/scripts/run_rank_llm.py:159]: I can add these two yes 👍🏼 I agree for the time being this is more clear

File: run_beir.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+
+# List of datasets
+datasets=("beir_cqadupstack-gis" "beir_cqadupstack-mathematica" "beir_cqadupstack-physics" "beir_cqadupstack-programmers" "beir_cqadupstack-stats" "beir_cqadupstack-tex" "beir_cqadupstack-unix" "beir_cqadupstack-webmasters" "beir_cqadupstack-wordpress" "beir_dbpedia-entity" "beir_fever" "beir_fiqa" "beir_hotpotqa" "beir_nfcorpus" "beir_nq" "beir_quora" "beir_robust04" "beir_scidocs" "beir_scifact" "beir_signal1m" "beir_trec-covid" "beir_trec-news" "beir_webis-touche2020")
+
+# List of GPUs
+GPUS=(0 1)
+
+# Function to run the command on a specific GPU
+run_on_gpu() {
```
Comment (sahel-sh) [inline run_beir.sh:10]: This relies on pyserini retriever, right? @lintool wants this retriever to be removed. Unless we decide to keep it, we should not add more dependencies to it.

File: src/rank_llm/rerank/rank_gpt.py [hunk 7]
```diff
@@ -56,9 +59,9 @@ def __init__(
             keys = [keys]
         if not keys:
             raise ValueError("Please provide OpenAI Keys.")
-        if prompt_mode not in [PromptMode.RANK_GPT, PromptMode.LRL]:
+        if prompt_mode not in [PromptMode.RANK_GPT, PromptMode.RANK_GPT_APEER, PromptMode.LRL]:
             raise ValueError(
-                f"unsupported prompt mode for GPT models: {prompt_mode}, expected {PromptMode.RANK_GPT} or {PromptMode.LRL}."
+                f"unsupported prompt mode for GPT models: {prompt_mode}, expected {PromptMode.RANK_GPT}, {PromptMode.RANK_GPT_APEER} or {PromptMode.LRL}."
```
Comment (sahel-sh) [inline src/rank_llm/rerank/rank_gpt.py:64]: why do we need this format? is it better than rank_gpt or different usage?

File: src/rank_llm/rerank/rank_gpt.py [hunk 8]
```diff
@@ -144,20 +148,35 @@ def run_llm(
     def _get_prefix_for_rank_gpt_prompt(
```
Comment (sahel-sh) [inline src/rank_llm/rerank/rank_gpt.py:148]: instead of introducing if/else in these functions please create separate helper functions:
e.g, _get_prefix_for_rank_gpt_apeer 

create_rank_gpt_prompt can call the proper functions based on prompt mode, this way your if/elses would be one-liners

File: src/rank_llm/rerank/rank_listwise_os_llm.py [hunk 9]
```diff
@@ -174,6 +224,27 @@ def create_prompt(
...
+
+        all_completed_prompts = []
+
+        with ThreadPoolExecutor() as executor:
+            for batch in tqdm(chunks(results, batch_size), desc="Processing batches"):
+                futures = [
+                    executor.submit(self.create_prompt, result, rank_start, rank_end)
+                    for result in batch
+                ]
+                completed_prompts = [future.result() for future in as_completed(futures)]
```
Comment (sahel-sh) [inline src/rank_llm/rerank/rank_listwise_os_llm.py:244]: IIUC, as completed returns the futures in order of completion rather than the original input or other. Should we use executor.map here?

File: src/rank_llm/scripts/convert_json_to_jsonl.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import argparse
+import json
+import os
+
+def convert_json_to_jsonl(input_file, output_file):
```
Comment (sahel-sh) [inline src/rank_llm/scripts/convert_json_to_jsonl.py:5]: why is this needed?

File: src/rank_llm/scripts/run_rank_llm.py (hunk 4 above)
Comment (sahel-sh) [reply src/rank_llm/scripts/run_rank_llm.py]: Less than Ideal since batch in rerank_batch has a different implication than batch in vllm-batched, but it is ok for now

File: src/rank_llm/scripts/convert_json_to_jsonl.py (hunk 10 above)
Comment (ronakice) [reply src/rank_llm/scripts/convert_json_to_jsonl.py:5]: Convert from unsupported JSON request format to newer JSONL request format.

File: run_beir.sh (hunk 6 above)
Comment (ronakice) [reply run_beir.sh:10]: Oops old stuff, removed

File: src/rank_llm/rerank/rank_gpt.py (hunk 7 above)
Comment (ronakice) [reply src/rank_llm/rerank/rank_gpt.py:64]: Another variant from here - https://arxiv.org/abs/2406.14449

Not too much work to add support, more people will stick to our framework for running that additional model too.

File: src/rank_llm/rerank/rank_listwise_os_llm.py (hunk 9 above)
Comment (ronakice) [reply src/rank_llm/rerank/rank_listwise_os_llm.py]: good catch, check now?

File: src/rank_llm/rerank/rank_gpt.py (hunk 8 above)
Comment (ronakice) [reply src/rank_llm/rerank/rank_gpt.py]: Done!
